README.md

# WP6RTools <img src="man/figures/WP6RTools.png" align="right" width="220"/>

WP6RTools provides a collection of tools to streamline the production
process of data analysis products. At the moment it includes two themes
for `{ggplot}` graphs and a color palette that emulates the OECD colors.

## Installation

Since this package is meant for internal use only, it will not be
distributed with CRAN. Simply type the following commands into your
R-Studio console to install the package from GitHub!

``` r
install.packages("remotes") # Lets you install packages from other sources than CRAN
remotes::install_github("BBieri/WP6RTools")
```

## OECD themes

As mentionned above, the package currently contains two, OECD-styled,
themes for `{ggplot}`. Both themes are essentially modified versions of
the `theme_ipsum()` theme of the
[`{hrbrthemes}`](https://github.com/hrbrmstr/hrbrthemes).

Let’s plot some things with the `theme_oecd()` light theme.

``` r
# Load required packages
library(WP6RTools) # Our package
library(dplyr) # Manipulation tools for dataframes. Part of the Tidyverse.
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(ggplot2) #
library(hrbrthemes)

# A scatter plot
ggplot(mtcars, aes(mpg, wt)) +
  geom_point() +
  labs(x = "Fuel efficiency (mpg)", y = "Weight (tons)",
       title = "Seminal ggplot2 scatterplot example",
       subtitle = "A plot that is only useful for demonstration purposes",
       caption = "Source: somewhere on the web.") + 
  theme_oecd()
```

<img src="man/figures/README-examplelight-1.png" width="100%" />

``` r
# A tidy histogram
count(mpg, class) %>%
  ggplot(aes(class, n)) +
   geom_col(fill = oecd_palette("oecd", 7, type = "continuous")) +
   geom_text(aes(label = n), nudge_y = 3) +
   labs(x = "Vehicle Category", y = "Number of Vehicles",
        title = "Seminal ggplot2 bar chart example",
        subtitle = "A plot that is only useful for demonstration purposes",
        caption = "Source: somewhere on the web") +
   theme_oecd(grid = "Y") +
   theme(axis.text.y = element_blank())
```

<img src="man/figures/README-examplelight-2.png" width="100%" />

Looking great! Let’s check out how these plots look like with the
`theme_oecd_dark()` dark theme.

``` r
# Load required packages
library(WP6RTools)
library(dplyr)
library(ggplot2)
library(hrbrthemes)

# A scatter plot
ggplot(mtcars, aes(mpg, wt)) +
  geom_point() +
  labs(x = "Fuel efficiency (mpg)", y = "Weight (tons)",
       title = "Seminal ggplot2 scatterplot example",
       subtitle = "A plot that is only useful for demonstration purposes",
       caption = "Source: somewhere on the web.") + 
  theme_oecd_dark()
```

<img src="man/figures/README-exampledark-1.png" width="100%" />

``` r
# A tidy histogram
count(mpg, class) %>% 
  mutate(pct=n/sum(n)) %>% # Create a new percentage column
  ggplot(aes(class, pct)) + # Create the graph
  geom_col() +
  scale_y_percent() +
  labs(y="Proportion", x="Vehicle category",
       title="Seminal ggplot2 column chart\nexample with percents",
       subtitle="A plot that is only useful for demonstration purposes",
       caption="Source: somewhere on the web.") + 
  theme_oecd_dark(grid="Y")
```

<img src="man/figures/README-exampledark-2.png" width="100%" />

## OECD palette generator

WP6Rtools also includes its very own palette generation tool based on
the included list `oecd_palettes`. You can generate both discrete color
palettes which will take the colors of the aforementionned list or
continuous palettes which will interpolate colors between the ones
specified in the `oecd_palettes` list.

``` r
# Discrete palette
discrete <- oecd_palette("oecd", 5, type = "discrete")
discrete
```

<img src="man/figures/README-oecdpalette-1.png" width="100%" />

``` r
# Continuous palette
continuous <- oecd_palette("oecd", 15, type = "continuous")
continuous
```

<img src="man/figures/README-oecdpalette-2.png" width="100%" />

## Miscellaneous OECD scraping functions

These functions are intended to produce quickly the list of OECD
countries, OECD Secretary Generals, and data about these two concepts.

``` r
head(oecd_members())
#> # A tibble: 6 x 8
#>   ID    Country  Application Negotiations Invitation Membership `Geographic loc~
#>   <chr> <chr>    <date>      <date>       <date>     <date>     <chr>           
#> 1 AUS   Austral~ NA          NA           NA         1971-06-07 Oceania         
#> 2 AUT   Austria  NA          NA           NA         1961-09-29 Europe          
#> 3 BEL   Belgium  NA          NA           NA         1961-09-13 Europe          
#> 4 CAN   Canada   NA          NA           NA         1961-04-10 North America   
#> 5 CHL   Chile    NA          2007-05-16   2009-12-15 2010-05-07 South America   
#> 6 COL   Colombia 2011-01-24  2013-05-30   2018-05-25 2020-04-28 South America   
#> # ... with 1 more variable: Notes <chr>
head(oecd_member_metrics())
#> # A tibble: 6 x 15
#>   ID    Country   `Area(km2)2017` `Population 2017` `GDP (PPP) (Intl. $)2017`
#>   <chr> <chr>               <dbl>             <dbl>                     <dbl>
#> 1 AUS   Australia         7741220          24598933             1192065505301
#> 2 AUT   Austria             83879           8809212              461582926400
#> 3 BEL   Belgium             30530          11372068              544041974958
#> 4 CAN   Canada            9984670          36708083             1714447151944
#> 5 CHL   Chile              756096          18054726              444777637169
#> 6 COL   Colombia          1141748          48901066              709420539907
#> # ... with 10 more variables: GDP (PPP)per capita(Intl. $)2017 <dbl>,
#> #   Incomeinequality 2008-2016(latest available) <dbl>, HDI2019 <dbl>,
#> #   FSI2019 <dbl>, RLI2020 <dbl>, CPI2019 <dbl>, IEF2020 <dbl>, GPI2019 <dbl>,
#> #   WPFI2019 <dbl>, DI2019 <dbl>
head(oecd_secretary())
#> # A tibble: 6 x 5
#>   ID    `Secretary-General`       Beg        End        `Country of origin`
#>   <chr> <chr>                     <date>     <date>     <chr>              
#> 1 1     Thorkil Kristensen        1961-09-30 1969-09-30 Denmark            
#> 2 2     Emiel van Lennep          1969-10-01 1984-09-30 Netherlands        
#> 3 3     Jean-Claude Paye          1984-10-01 1994-09-30 France             
#> 4 —     Staffan Sohlman (interim) 1994-10-01 1994-11-01 Sweden             
#> 5 3     Jean-Claude Paye          1994-11-01 1996-05-31 France             
#> 6 4     Donald Johnston           1996-06-01 2006-05-31 Canada
```

## Issues, Improvements and Questions

Have you experienced an issue while using the package or have a question
about it? Open an [issue](https://github.com/BBieri/WP6RTools/issues).
Do you want to suggest improvement? Then open up a
[PR](https://github.com/BBieri/WP6RTools/pulls).

## Final Notes

Note that the present R-Package is in no way related to or endorsed by
the OECD.

Please note that the WP6RTools project is released with a [Contributor
Code of
Conduct](https://bbieri.github.io/WP6RTools/CODE_OF_CONDUCT.html). By
contributing to this project, you agree to abide by its terms.


NEWS.md
# WP6RTools 0.0.2

- Added `oecd_secretary()` a function scraping the table of Secretary Generals
of the OECD from Wikipedia
- Added `oecd_members()` a function scraping the table of member countries
of the OECD from Wikipedia
- Added `oecd_member_metrics()` a function scraping the table of member
countries of the OECD and their socioeconomic characteristics from Wikipedia

# WP6RTools 0.0.1

- Added `theme_oecd()` and `theme_oecd_dark()`, two themes for `{ggplot}` graphs
- Added `oecd_palettes()` a list of oecd palettes
- Added `oecd_palette()` a function to generate discrete and continuous palettes
based on the OECD colors

# WP6RTools 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(WP6RTools)

test_check("WP6RTools")


tests/testthat/test-OECD_misc.R
# Run tests
test_that("oecd_members has the correct countries", {
  expect_equal(oecd_members()[[2]], c('Australia', 'Austria', 'Belgium',
                                      'Canada', 'Chile', 'Colombia',
                                      'Costa Rica', 'Czech Republic',
                                      'Denmark', 'Estonia', 'Finland',
                                      'France', 'Germany', 'Greece',
                                      'Hungary', 'Iceland', 'Ireland',
                                      'Israel', 'Italy', 'Japan', 'South Korea',
                                      'Latvia', 'Lithuania', 'Luxembourg',
                                      'Mexico', 'Netherlands', 'New Zealand',
                                      'Norway', 'Poland', 'Portugal',
                                      'Slovakia', 'Slovenia', 'Spain',
                                      'Sweden', 'Switzerland', 'Turkey',
                                      'United Kingdom', 'United States'))
           })
test_that("oecd_member_metrics has the correct countries", {
  expect_equal(oecd_member_metrics()[[2]], c('Australia', 'Austria', 'Belgium',
                                      'Canada', 'Chile', 'Colombia',
                                      'Costa Rica', 'Czech Republic',
                                      'Denmark', 'Estonia', 'Finland',
                                      'France', 'Germany', 'Greece',
                                      'Hungary', 'Iceland', 'Ireland',
                                      'Israel', 'Italy', 'Japan', 'South Korea',
                                      'Latvia', 'Lithuania', 'Luxembourg',
                                      'Mexico', 'Netherlands', 'New Zealand',
                                      'Norway', 'Poland', 'Portugal',
                                      'Slovakia', 'Slovenia', 'Spain',
                                      'Sweden', 'Switzerland', 'Turkey',
                                      'United Kingdom', 'United States',
                                      'OECDbc'))
})
test_that("oecd_secretary has the correct number of secretaries", {
  expect_equal(length(oecd_secretary()[[2]]), 8)
})


tests/testthat/test-OECD_theme.R
# Create the plot to test
p <- dplyr::count(ggplot2::mpg, class) %>%
  ggplot2::ggplot(ggplot2::aes(class, n)) +
  ggplot2::geom_col(fill = oecd_palette("oecd", 7, type = "continuous")) +
  ggplot2::geom_text(ggplot2::aes(label = n), nudge_y = 3) +
  ggplot2::labs(x = "Vehicle Category", y = "Number of Vehicles",
                title = "Seminal ggplot2 bar chart example",
                subtitle = "A plot that is only useful for demonstration purposes",
                caption = "Source: somewhere on the web") +
  theme_oecd(grid = "Y") +
  ggplot2::theme(axis.text.y = element_blank())
# Run tests on plots
test_that("Colours of the light plot theme are correct", {
  expect_equal(p[["theme"]][["line"]][["colour"]], "black")
  expect_equal(p[["theme"]][["rect"]][["colour"]], "black")
  expect_equal(p[["theme"]][["text"]][["colour"]], "black")
           })


R/OECD_palette.R
#' Complete list of palettes
#'
#' Discrete color palette based on OECD colors.
#'
#' @export

oecd_palettes <- list(
  "oecd" = c(
    "oecd_gray" = "#343a40",
    "oecd_blue" = "#264653",
    "oecd_turquoise" = "#2a9d8f",
    "oecd_yellow" = "#e9c46a",
    "oecd_orange" = "#f4a261",
    "oecd_red" = "#e76f51"
  )

)

#' An OECD palette generator
#'
#' This is a small palette generator for OECD graphs that generates color vectors
#' to be used in graphs
#'
#' @param n Number of colors desired. If omitted, uses all colours.
#' @param name Name of desired palette. Current choice is:
#' \code{oecd}
#' @param type Either "continuous" or "discrete". Use continuous if you want
#'   to automatically interpolate between colours.
#' @importFrom graphics rect par image text
#' @return A vector of colours.
#' @source Adapted from https://github.com/karthik/wesanderson/blob/master/R/colors.R
#' @export
#' @keywords colors
#' @examples
#' oecd_palette("oecd")
#'
#' # If you need more colours than normally found in a palette, you
#' # can use a continuous palette to interpolate between existing
#' # colours
#' pal <- oecd_palette(21, name = "oecd", type = "continuous")
#' image(volcano, col = pal)
oecd_palette <- function(name, n, type = c("discrete", "continuous")) {
  type <- match.arg(type)
  pal <- oecd_palettes[[name]]
  if (is.null(pal))
    stop("Palette not found.")
  if (missing(n)) {
    n <- length(pal)
  }
  if (type == "discrete" && n > length(pal)) {
    stop("Number of requested colors greater than what palette can offer")
  }
  out <- switch(type,
                continuous = grDevices::colorRampPalette(pal)(n),
                discrete = pal[1:n]
  )
  structure(out, class = "palette", name = name)
}

#' @export
#' @importFrom graphics rect par image text
#' @importFrom grDevices rgb
print.palette <- function(x, ...) {
  n <- length(x)
  old <- par(mar = c(0.5, 0.5, 0.5, 0.5))
  on.exit(par(old))
  image(1:n, 1, as.matrix(1:n), col = x,
        ylab = "", xaxt = "n", yaxt = "n", bty = "n")
  rect(0, 0.9, n + 1, 1.1, col = rgb(1, 1, 1, 0.8), border = NA)
  text((n + 1) / 2, 1, labels = attr(x, "name"), cex = 1, family = "serif")
}


R/OECD_theme.R
#' A ggplot2 theme for OECD style plots. Adapted from Hrbrmstr's ipsum theme.
#'
#' @importFrom ggplot2 theme margin element_text element_blank element_line element_rect
#' @param base_family,base_size base font family and size
#' @param main_col main color of the plot. Colors points, histogram bars, etc.
#' @param plot_title_family,plot_title_face,plot_title_size,plot_title_margin plot title family, face, size and margi
#' @param subtitle_family,subtitle_face,subtitle_size plot subtitle family, face and size
#' @param subtitle_margin plot subtitle margin bottom (single numeric value)
#' @param strip_text_family,strip_text_face,strip_text_size facet label font family, face and size
#' @param caption_family,caption_face,caption_size,caption_margin plot caption family, face, size and margin
#' @param axis_title_family,axis_title_face,axis_title_size axis title font family, face and size
#' @param axis_title_just axis title font justification, one of `[blmcrt]`
#' @param plot_margin plot margin (specify with `ggplot2::margin()`)
#' @param grid_col,axis_col grid & axis colors; both default to `#cccccc`
#' @param grid panel grid (`TRUE`, `FALSE`, or a combination of `X`, `x`, `Y`, `y`)
#' @param axis_text_size font size of axis text
#' @param axis add x or y axes? `TRUE`, `FALSE`, "`xy`"
#' @param ticks ticks if `TRUE` add ticks
#' @return Themes the current ggplot to current IHEID guidelines.
#' @examples
#' library(ggplot2)
#' library(dplyr)
#'
#' # seminal scatterplot
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
#'        title="Seminal ggplot2 scatterplot example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Brought to you by the letter 'g'") +
#'   theme_oecd()
#'
#' # seminal bar chart
#'
#' count(mpg, class) %>%
#'  ggplot(aes(class, n)) +
#'   geom_col(fill = oecd_palette("oecd", 7, type = "continuous")) +
#'   geom_text(aes(label=n), nudge_y=3) +
#'   labs(x="Vehicle Category", y="Number of Vehicles",
#'        title="Seminal ggplot2 bar chart example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Source: somewhere on the web") +
#'   theme_oecd(grid="Y") +
#'   theme(axis.text.y=element_blank())
#' @export

theme_oecd <- function(base_family = "sans",
                        base_size = 11.5,
                        main_col = "grey35",
                        plot_title_family = base_family,
                        plot_title_size = 18,
                        plot_title_face = "bold",
                        plot_title_margin = 10,
                        subtitle_family = base_family,
                        subtitle_size = 12,
                        subtitle_face = "plain",
                        subtitle_margin = 15,
                        strip_text_family = base_family,
                        strip_text_size = 12,
                        strip_text_face = "plain",
                        caption_family = base_family,
                        caption_size = 9,
                        caption_face = "italic",
                        caption_margin = 10,
                        axis_text_size = base_size,
                        axis_title_family = subtitle_family,
                        axis_title_size = 9,
                        axis_title_face = "plain",
                        axis_title_just = "rt",
                        plot_margin = margin(30, 30, 30, 30),
                        grid_col = "grey85",
                        grid = TRUE,
                        axis_col = "grey35",
                        axis = FALSE,
                        ticks = FALSE)
{
  # Redefine some styles. Point colors, histogram colors etc...
  ggplot2::update_geom_defaults("point", list(colour = main_col))
  ggplot2::update_geom_defaults("line", list(colour = main_col))
  ggplot2::update_geom_defaults("area", list(colour = main_col,
                                             fill = main_col))
  ggplot2::update_geom_defaults("rect", list(colour = main_col,
                                             fill = main_col))
  ggplot2::update_geom_defaults("density", list(colour = main_col,
                                                fill = main_col))
  ggplot2::update_geom_defaults("bar", list(colour = main_col,
                                            fill = main_col))
  ggplot2::update_geom_defaults("col", list(colour = main_col,
                                            fill = main_col))
  ggplot2::update_geom_defaults("text", list(colour = "gray85"))
  # Apply minimal theme from ggplot
  ret <- ggplot2::theme_minimal(base_family = base_family,
                                base_size = base_size)
  ret <- ret + theme(legend.background = element_blank())
  ret <- ret + theme(legend.key = element_blank())
  if (inherits(grid, "character") | grid == TRUE) {
    ret <- ret + theme(panel.grid = element_line(color = grid_col,
                                                 size = 0.2))
    ret <- ret + theme(panel.grid.major = element_line(color = grid_col,
                                                       size = 0.2))
    ret <- ret + theme(panel.grid.minor = element_line(color = grid_col,
                                                       size = 0.15))
    if (inherits(grid, "character")) {
      if (regexpr("X", grid)[1] < 0)
        ret <- ret + theme(panel.grid.major.x = element_blank())
      if (regexpr("Y", grid)[1] < 0)
        ret <- ret + theme(panel.grid.major.y = element_blank())
      if (regexpr("x", grid)[1] < 0)
        ret <- ret + theme(panel.grid.minor.x = element_blank())
      if (regexpr("y", grid)[1] < 0)
        ret <- ret + theme(panel.grid.minor.y = element_blank())
    }
  }
  else {
    ret <- ret + theme(panel.grid = element_blank())
  }
  if (inherits(axis, "character") | axis == TRUE) {
    ret <- ret + theme(axis.line = element_line(color = "#2b2b2b",
                                                size = 0.15))
    if (inherits(axis, "character")) {
      axis <- tolower(axis)
      if (regexpr("x", axis)[1] < 0) {
        ret <- ret + theme(axis.line.x = element_blank())
      }
      else {
        ret <- ret + theme(axis.line.x = element_line(color = axis_col,
                                                      size = 0.15))
      }
      if (regexpr("y", axis)[1] < 0) {
        ret <- ret + theme(axis.line.y = element_blank())
      }
      else {
        ret <- ret + theme(axis.line.y = element_line(color = axis_col,
                                                      size = 0.15))
      }
    }
    else {
      ret <- ret + theme(axis.line.x = element_line(color = axis_col,
                                                    size = 0.15))
      ret <- ret + theme(axis.line.y = element_line(color = axis_col,
                                                    size = 0.15))
    }
  }
  else {
    ret <- ret + theme(axis.line = element_blank())
  }
  if (!ticks) {
    ret <- ret + theme(axis.ticks = element_blank())
    ret <- ret + theme(axis.ticks.x = element_blank())
    ret <- ret + theme(axis.ticks.y = element_blank())
  }
  else {
    ret <- ret + theme(axis.ticks = element_line(size = 0.15))
    ret <- ret + theme(axis.ticks.x = element_line(size = 0.15))
    ret <- ret + theme(axis.ticks.y = element_line(size = 0.15))
    ret <- ret + theme(axis.ticks.length = grid::unit(5,
                                                      "pt"))
  }
  xj <- switch(tolower(substr(axis_title_just, 1, 1)), b = 0,
               l = 0, m = 0.5, c = 0.5, r = 1, t = 1)
  yj <- switch(tolower(substr(axis_title_just, 2, 2)), b = 0,
               l = 0, m = 0.5, c = 0.5, r = 1, t = 1)
  ret <- ret + theme(axis.text.x = element_text(size = axis_text_size,
                                                margin = margin(t = 0)))
  ret <- ret + theme(axis.text.y = element_text(size = axis_text_size,
                                                margin = margin(r = 0)))
  ret <- ret + theme(axis.title = element_text(size = axis_title_size,
                                               family = axis_title_family))
  ret <- ret + theme(axis.title.x = element_text(hjust = xj,
                                                 size = axis_title_size, family = axis_title_family,
                                                 face = axis_title_face))
  ret <- ret + theme(axis.title.y = element_text(hjust = yj,
                                                 size = axis_title_size, family = axis_title_family,
                                                 face = axis_title_face))
  ret <- ret + theme(axis.title.y.right = element_text(hjust = yj,
                                                       size = axis_title_size, angle = 90, family = axis_title_family,
                                                       face = axis_title_face))
  ret <- ret + theme(strip.text = element_text(hjust = 0,
                                               size = strip_text_size, face = strip_text_face, family = strip_text_family))
  ret <- ret + theme(panel.spacing = grid::unit(2, "lines"))
  ret <- ret + theme(plot.title = element_text(hjust = 0,
                                               size = plot_title_size, margin = margin(b = plot_title_margin),
                                               family = plot_title_family, face = plot_title_face))
  ret <- ret + theme(plot.subtitle = element_text(hjust = 0,
                                                  size = subtitle_size, margin = margin(b = subtitle_margin),
                                                  family = subtitle_family, face = subtitle_face))
  ret <- ret + theme(plot.caption = element_text(hjust = 1,
                                                 size = caption_size, margin = margin(t = caption_margin),
                                                 family = caption_family, face = caption_face))
  ret <- ret + theme(plot.margin = plot_margin)
  ret
}

#' A ggplot2 theme for OECD style plots. Adapted from Hrbrmstr's ipsum theme.
#'
#' @importFrom ggplot2 theme margin element_text element_blank element_line element_rect
#' @param base_family,base_size base font family and size
#' @param main_col main color of the plot. Colors points, histogram bars, etc.
#' @param plot_title_family,plot_title_face,plot_title_size,plot_title_margin plot title family, face, size and margi
#' @param subtitle_family,subtitle_face,subtitle_size,subtitle_col plot subtitle family, face, size, and color
#' @param subtitle_margin plot subtitle margin bottom (single numeric value)
#' @param strip_text_family,strip_text_face,strip_text_size facet label font family, face and size
#' @param caption_family,caption_face,caption_size,caption_margin plot caption family, face, size and margin
#' @param axis_title_family,axis_title_face,axis_title_size axis title font family, face and size
#' @param axis_title_just axis title font justification, one of `[blmcrt]`
#' @param plot_margin plot margin (specify with `ggplot2::margin()`)
#' @param grid panel grid (`TRUE`, `FALSE`, or a combination of `X`, `x`, `Y`, `y`)
#' @param axis_text_size font size of axis text
#' @param axis add x or y axes? `TRUE`, `FALSE`, "`xy`"
#' @param ticks ticks if `TRUE` add ticks
#' @return Themes the current ggplot to current IHEID guidelines.
#' @examples
#' library(ggplot2)
#' library(dplyr)
#'
#' # Seminal scatterplot
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
#'        title="Seminal ggplot2 scatterplot example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Brought to you by the letter 'g'") +
#'   theme_oecd_dark()
#'
#' # Seminal bar chart
#'
#' count(mpg, class) %>%
#'  ggplot(aes(class, n)) +
#'    geom_col(fill = oecd_palette("oecd", 21, type = "continuous")[15:21]) +
#'    geom_text(aes(label=n), nudge_y=3) +
#'    labs(x="Vehicle Category", y="Number of Vehicles",
#'         title="Seminal ggplot2 bar chart example",
#'         subtitle="A plot that is only useful for demonstration purposes",
#'         caption="Source: somewhere on the web") +
#'    theme_oecd(grid="Y") +
#'    theme(axis.text.y=element_blank())
#' @export

theme_oecd_dark <- function(base_family = "sans",
                             base_size = 11.5,
                             main_col = "#EBBB67",
                             plot_title_family = base_family,
                             plot_title_size = 18,
                             plot_title_face = "bold",
                             plot_title_margin = 10,
                             subtitle_family = "sans",
                             subtitle_size = 13,
                             subtitle_face = "plain",
                             subtitle_margin = 15,
                             subtitle_col = "gray85",
                             strip_text_family = base_family,
                             strip_text_size = 12,
                             strip_text_face = "plain",
                             caption_family = "sans",
                             caption_size = 9,
                             caption_face = "plain",
                             caption_margin = 10,
                             axis_text_size = base_size,
                             axis_title_family = base_family,
                             axis_title_size = 9,
                             axis_title_face = "plain",
                             axis_title_just = "rt",
                             plot_margin = margin(30, 30, 30, 30),
                             grid = TRUE,
                             axis = FALSE,
                             ticks = FALSE)
{
  grid_col <- axis_col <- "#464950"
  ggplot2::update_geom_defaults("point", list(colour = main_col))
  ggplot2::update_geom_defaults("line", list(colour = main_col))
  ggplot2::update_geom_defaults("area", list(colour = main_col,
                                             fill = main_col))
  ggplot2::update_geom_defaults("rect", list(colour = main_col,
                                             fill = main_col))
  ggplot2::update_geom_defaults("density", list(colour = main_col,
                                                fill = main_col))
  ggplot2::update_geom_defaults("bar", list(colour = main_col,
                                            fill = main_col))
  ggplot2::update_geom_defaults("col", list(colour = main_col,
                                            fill = main_col))
  ggplot2::update_geom_defaults("text", list(colour = "gray85"))
  ret <- ggplot2::theme_minimal(base_family = base_family,
                                base_size = base_size)
  ret <- ret + theme(legend.background = element_blank())
  ret <- ret + theme(legend.key = element_blank())
  if (inherits(grid, "character") | grid == TRUE) {
    ret <- ret + theme(panel.grid = element_line(color = grid_col,
                                                 size = 0.2))
    ret <- ret + theme(panel.grid.major = element_line(color = grid_col,
                                                       size = 0.2))
    ret <- ret + theme(panel.grid.minor = element_line(color = grid_col,
                                                       size = 0.15))
    if (inherits(grid, "character")) {
      if (regexpr("X", grid)[1] < 0)
        ret <- ret + theme(panel.grid.major.x = element_blank())
      if (regexpr("Y", grid)[1] < 0)
        ret <- ret + theme(panel.grid.major.y = element_blank())
      if (regexpr("x", grid)[1] < 0)
        ret <- ret + theme(panel.grid.minor.x = element_blank())
      if (regexpr("y", grid)[1] < 0)
        ret <- ret + theme(panel.grid.minor.y = element_blank())
    }
  }
  else {
    ret <- ret + theme(panel.grid = element_blank())
  }
  if (inherits(axis, "character") | axis == TRUE) {
    ret <- ret + theme(axis.line = element_line(color = "white",
                                                size = 0.15))
    if (inherits(axis, "character")) {
      axis <- tolower(axis)
      if (regexpr("x", axis)[1] < 0) {
        ret <- ret + theme(axis.line.x = element_blank())
      }
      else {
        ret <- ret + theme(axis.line.x = element_line(color = axis_col,
                                                      size = 0.15))
      }
      if (regexpr("y", axis)[1] < 0) {
        ret <- ret + theme(axis.line.y = element_blank())
      }
      else {
        ret <- ret + theme(axis.line.y = element_line(color = axis_col,
                                                      size = 0.15))
      }
    }
    else {
      ret <- ret + theme(axis.line.x = element_line(color = axis_col,
                                                    size = 0.15))
      ret <- ret + theme(axis.line.y = element_line(color = axis_col,
                                                    size = 0.15))
    }
  }
  else {
    ret <- ret + theme(axis.line = element_blank())
  }
  if (!ticks) {
    ret <- ret + theme(axis.ticks = element_blank())
    ret <- ret + theme(axis.ticks.x = element_blank())
    ret <- ret + theme(axis.ticks.y = element_blank())
  }
  else {
    ret <- ret + theme(axis.ticks = element_line(size = 0.15))
    ret <- ret + theme(axis.ticks.x = element_line(size = 0.15))
    ret <- ret + theme(axis.ticks.y = element_line(size = 0.15))
    ret <- ret + theme(axis.ticks.length = grid::unit(5,
                                                      "pt"))
  }
  xj <- switch(tolower(substr(axis_title_just, 1, 1)), b = 0,
               l = 0, m = 0.5, c = 0.5, r = 1, t = 1)
  yj <- switch(tolower(substr(axis_title_just, 2, 2)), b = 0,
               l = 0, m = 0.5, c = 0.5, r = 1, t = 1)
  ret <- ret + theme(axis.text.x = element_text(size = axis_text_size,
                                                margin = margin(t = 0)))
  ret <- ret + theme(axis.text.y = element_text(size = axis_text_size,
                                                margin = margin(r = 0)))
  ret <- ret + theme(axis.title = element_text(size = axis_title_size,
                                               family = axis_title_family))
  ret <- ret + theme(axis.title.x = element_text(hjust = xj,
                                                 size = axis_title_size, family = axis_title_family,
                                                 face = axis_title_face))
  ret <- ret + theme(axis.title.y = element_text(hjust = yj,
                                                 size = axis_title_size, family = axis_title_family,
                                                 face = axis_title_face))
  ret <- ret + theme(strip.text = element_text(hjust = 0,
                                               size = strip_text_size, color = subtitle_col, face = strip_text_face,
                                               family = strip_text_family))
  ret <- ret + theme(panel.spacing = grid::unit(2, "lines"))
  ret <- ret + theme(plot.title = element_text(hjust = 0,
                                               size = plot_title_size, margin = margin(b = plot_title_margin),
                                               family = plot_title_family, face = plot_title_face))
  ret <- ret + theme(plot.subtitle = element_text(hjust = 0,
                                                  size = subtitle_size, color = subtitle_col, margin = margin(b = subtitle_margin),
                                                  family = subtitle_family, face = subtitle_face))
  ret <- ret + theme(plot.caption = element_text(hjust = 1,
                                                 size = caption_size, margin = margin(t = caption_margin),
                                                 family = caption_family, face = caption_face))
  ret <- ret + theme(plot.margin = plot_margin)
  bkgrnd <- oecd_palettes[["oecd"]][["oecd_gray"]] # Background color
  fgrnd <- "#617a89" # Foreground color
  ret <- ret + theme(rect = element_rect(fill = bkgrnd, color = bkgrnd)) +
    theme(plot.background = element_rect(fill = bkgrnd,
                                         color = bkgrnd)) + theme(panel.background = element_rect(fill = bkgrnd,
                                                                                                  color = bkgrnd)) + theme(rect = element_rect(fill = bkgrnd,
                                                                                                                                               color = bkgrnd)) + theme(text = element_text(color = "#929299")) +
    theme(axis.text = element_text(color = "#929299")) +
    theme(title = element_text(color = "#929299")) + theme(plot.title = element_text(color = "white")) +
    theme(plot.subtitle = element_text(color = "#929299")) +
    theme(plot.caption = element_text(color = "#929299")) +
    theme(line = element_line(color = grid_col)) + theme(axis.ticks = element_line(color = grid_col))
}
